service/controller/v12/resource/instance/create.go

```diff
@@ -613,14 +630,26 @@ func firstInstanceToReimage(customObject providerv1alpha1.AzureConfig, list []co
 			return nil, microerror.Mask(err)
 		}
 		instanceVersion, ok := versionValue[instanceName]
+		// current version unavailable, skip this instance
 		if !ok {
 			continue
 		}
-		if desiredVersion == instanceVersion {
-			continue
+		// version is changed
+		if desiredVersion != instanceVersion {
+			return &v, nil
 		}
 
-		return &v, nil
+		for _, disk := range *v.StorageProfile.DataDisks {
+			desiredDiskSize, ok := desiredDiskSizes[*disk.Name]
+			if !ok {
+				// this disk has no target size specified, ignore it
+				continue
+			}
+			// desired disk size is changed
+			if *disk.DiskSizeGB != desiredDiskSize {
+				return &v, nil
+			}
+		}
```

Comment:
I'd maybe change this to be something like following:
```go
		for _, disk := range *v.StorageProfile.DataDisks {
			switch(*disk.Name) {
			case key.DockerVolumeDiskName:
				if *disk.DiskSizeGB != key.DockerVolumeSizeGB(customObject) {
					return &v, nil
				}
			case key.KubeletVolumeDiskName:
				if *disk.DiskSizeGB != key.KubeletVolumeSizeGB(customObject) {
					return &v, nil
				}
			}
		}
```

For that you need to add corresponding key functions (similar to this: https://github.com/giantswarm/aws-operator/blob/master/service/controller/key/machine_deployment.go#L16-L18) and constants to match for disk name.

`azure-operator` key functions are here: https://github.com/giantswarm/azure-operator/blob/master/service/controller/v12/key/key.go